broker/src/task_manager.rs

```diff
@@ -0,0 +1,329 @@
+                                return Err(TaskManagerError::Gone);
+                            }
+                        },
+                        Err(e) => {
+                            warn!("new_results channel lagged: {e}");
+                            return Err(TaskManagerError::BroadcastBufferOverflow);
+                        }
+                    }
+                },
+            }
+        }
+
+        Ok(self.get(task_id).map_err(|_| TaskManagerError::Gone)?)
+    }
+
+    pub fn stream_results<'a>(
+        &'a self,
+        task_id: &'a MsgId,
+        block: &'a HowLongToBlock,
+        filter: impl Fn(&T::Result) -> bool + 'a
+    ) -> Result<impl Stream<Item = Result<Event, TaskManagerError>> + 'a, TaskManagerError>
+        where T::Result: Serialize
+    {
+        let task = self.get(task_id)?;
+        Ok(async_stream::stream! {
+            let max_elements = block.wait_count.unwrap_or(u16::MAX) as usize;
+            let wait_until = Instant::now() + block.wait_time.unwrap_or(Duration::from_secs(600));
+            let ready_results = task.msg
+                .get_results()
+                .values()
+                .filter(|result| filter(result));
+            let mut num_of_results = 0;
+            for res in ready_results {
+                yield Ok(to_event(res, SseEventType::NewResult));
+                num_of_results += 1;
+                if num_of_results >= max_elements {
+                    break;
+                }
+            }
+            let mut new_results = self
+                .new_results
+                .get(task_id)
+                .ok_or(TaskManagerError::NotFound)?
+                .subscribe();
+            while num_of_results < max_elements && Instant::now() < wait_until {
+                tokio::select! {
+                    _ = tokio::time::sleep_until(wait_until) => {
+                        break;
+                    },
+                    result = new_results.recv() => {
+                        match result {
+                            Ok(key) => {
+                                if let Ok(task) = self.get(task_id) {
+                                    let new_result = &task.msg.get_results()[&key];
+                                    if filter(new_result) {
+                                        num_of_results += 1;
+                                        yield Ok(to_event(new_result, SseEventType::NewResult));
+                                    };
+                                } else {
+                                    yield Err(TaskManagerError::Gone);
+                                }
+                            },
+                            Err(e) => {
+                                warn!("new_results channel lagged: {e}");
+                                yield Err(TaskManagerError::BroadcastBufferOverflow);
+                            }
+                        }
+                    },
+                }
+            }
+        })
+    }
```

Comment:
That's some deep nesting. Maybe refactoring this is useful for code maintainability.

Reply:
I agree but I am honestly not sure how or what to extract here because the yields need to be inside the `async_stream!` macro and most of the nesting is because of the `tokio::select!`.